src/frequenz/client/dispatch/types.py

```diff
@@ -21,16 +32,180 @@
-
-# pylint: enable=no-name-in-module
 from frequenz.client.common.microgrid.components import ComponentCategory
 
 from .recurrence import Frequency, RecurrenceRule, Weekday
 
-TargetComponents = list[int] | list[ComponentCategory]
-"""One or more target components specifying which components a dispatch targets.
 
-It can be a list of component IDs or a list of categories.
+class EvChargerType(IntEnum):
+    """Enum representing the type of EV charger."""
+
+    UNSPECIFIED = PBEvChargerType.EV_CHARGER_TYPE_UNSPECIFIED
+    """Unspecified type of EV charger."""
+
+    AC = PBEvChargerType.EV_CHARGER_TYPE_AC
+    """AC EV charger."""
+
+    DC = PBEvChargerType.EV_CHARGER_TYPE_DC
+    """DC EV charger."""
+
+    HYBRID = PBEvChargerType.EV_CHARGER_TYPE_HYBRID
+    """Hybrid EV charger."""
+
+
+class BatteryType(IntEnum):
+    """Enum representing the type of battery."""
+
+    UNSPECIFIED = PBBatteryType.BATTERY_TYPE_UNSPECIFIED
+    """Unspecified type of battery."""
+
+    LI_ION = PBBatteryType.BATTERY_TYPE_LI_ION
+    """Lithium-ion battery."""
+
+    NA_ION = PBBatteryType.BATTERY_TYPE_NA_ION
+    """Sodium-ion battery."""
+
+
+class InverterType(IntEnum):
+    """Enum representing the type of inverter."""
+
+    UNSPECIFIED = PBInverterType.INVERTER_TYPE_UNSPECIFIED
+    """Unspecified type of inverter."""
+
+    BATTERY = PBInverterType.INVERTER_TYPE_BATTERY
+    """Battery inverter."""
+
+    SOLAR = PBInverterType.INVERTER_TYPE_SOLAR
+    """Solar inverter."""
+
+    HYBRID = PBInverterType.INVERTER_TYPE_HYBRID
+    """Hybrid inverter."""
+
+
+@dataclass(frozen=True)
+class TargetCategory:
+    """Represents a category and optionally a type."""
+
+    target: ComponentCategory | BatteryType | EvChargerType | InverterType
+    """The target category of the dispatch.
+
+    Implicitly derived from the types.
+    """
+
+    @property
+    def category(self) -> ComponentCategory:
+        """Get the category of the target.
+
+        Returns:
+            The category of the target.
+        """
+        match self.target:
+            case ComponentCategory():
+                return self.target
+            case BatteryType():
+                return ComponentCategory.BATTERY
+            case EvChargerType():
+                return ComponentCategory.EV_CHARGER
+            case InverterType():
+                return ComponentCategory.INVERTER
+
+    @property
+    def type(self) -> BatteryType | EvChargerType | InverterType | None:
+        """Get the type of the category.
+
+        Returns:
+            The type of the category.
+        """
+        match self.target:
+            case BatteryType() | EvChargerType() | InverterType():
+                return self.target
+            case _:
+                return None
+
+
+class TargetIds(frozenset[int]):
+    """A set of target component IDs.
+
+    This is a frozen set, so it is immutable.
+    """
+
+    def __new__(cls, *ids: SupportsInt) -> "TargetIds":
```

Comment:
`from typing import Self`:

```suggestion
    def __new__(cls, *ids: SupportsInt) -> Self:
```

Probably it will not be subclasses, but if it is, the type should be the type of the sub-class, not always this one.